### ISSUE TITLE
Push to non-existent upstream branch messes up

### ISSUE THREAD

**Issue by ingydotnet** [ingydotnet/git-subrepo#106]
Here is a replication strategy from memory:

* Clone ingydotnet/git-subrepo and checkout tag 0.3.0
* Make a fork of ingydotnet/test-more-bash
* Run this command:

    git subrepo push ext/test-more-bash \
        --remote=git@github.com/you/test-more-bash \
        --branch=new-test-branch

* You should a new branch on your fork with 40 commits
* Run subrepo push command again
  * You see 42 commits
* Run again
  * Command complains
* Run again
  * Command says up to date

Basically you can override the remote branch on a push, but if the branch
doesn't exist, it does it wrong.

**Comment by ingydotnet** [comment]
Here's a quick look at what I see: https://gist.github.com/anonymous/b8e81fcbbfdc4348466f

**Comment by grimmySwe** [comment]
@ingydotnet Let's see if get this correctly, summarizing the gist:
```
> git subrepo push -vd ext/test-more-bash/ --branch=issue106     # (1)
* Pushing to new upstream: git@github.com:ingydotnet/test-more-bash.git (issue106).
Subrepo 'ext/test-more-bash' pushed to 'git@github.com:ingydotnet/test-more-bash.git' (issue106).
> git subrepo push -vd ext/test-more-bash/ --branch=issue106     # (2)
Subrepo 'ext/test-more-bash' pushed to 'git@github.com:ingydotnet/test-more-bash.git' (issue106).
> git subrepo push -vd ext/test-more-bash/ --branch=issue106     # (3)
git-subrepo: Local branch is not updated, perform pull or use '--force' to always trust local branch in conflicts
> git subrepo push -vd ext/test-more-bash/ --branch=issue106     # (4)
git-subrepo: Local branch is not updated, perform pull or use '--force' to always trust local branch in conflicts
```
First (1) push will create the new branch. Second (2) should though not push new stuff, it should simply detect that there is no difference. 
(3) and (4) as as expected in this case as (2) has changed something upstream and you haven't pulled the changes back.

I'll see if I can create a simple test case for this.

**Comment by grimmySwe** [comment]
This test case seems to show the problem:
```
source test/setup

use Test::More

clone-foo-and-bar

# Make various changes to the repos for testing subrepo push:
(
  # In the main repo:
  cd $OWNER/foo

  # Clone the subrepo into a subdir
  git subrepo clone ../../../$UPSTREAM/bar

  # Make a commit:
  add-new-files bar/FooBar
) &> /dev/null || die

# Do the subrepo push to another branch:
{
  message="$(
    cd $OWNER/foo
    git subrepo push bar --branch newbar
  )"

  # Test the output:
  is "$message" \
    "Subrepo 'bar' pushed to '../../../tmp/upstream/bar' (newbar)." \
    'push message is correct'
}

# Do the subrepo push to another branch again:
{
  message="$(
    cd $OWNER/foo
    git subrepo push bar --branch newbar
  )"

  # Test the output:
  is "$message" \
    "git-subrepo: Local branch is not updated, perform pull or use
'--force' to always trust local branch in conflicts" \
      'push message is correct'
}


# Pull the changes from UPSTREAM/bar in OWNER/bar
(
  cd $OWNER/bar
  git fetch
  git checkout newbar
  git rebase -p
) &> /dev/null || die

test-exists \
  "$OWNER/bar/FooBar" \

done_testing

teardown
```

```
$ prove test/subrepo-push-new-branch.t 
test/subrepo-push-new-branch.t .. 2/? #   Failed test 'push message is correct'
#   at test/subrepo-push-new-branch.t line 48.
--- /tmp/want-32536     2015-12-03 08:38:00.340933526 +0100
+++ /tmp/got-32536      2015-12-03 08:38:00.340933526 +0100
@@ -1,2 +1 @@
-git-subrepo: Local branch is not updated, perform pull or use
-'--force' to always trust local branch in conflicts
+Subrepo 'bar' pushed to '../../../tmp/upstream/bar' (newbar).
# Tests were run but no plan was declared.
test/subrepo-push-new-branch.t .. Failed 1/2 subtests 

Test Summary Report
-------------------
test/subrepo-push-new-branch.t (Wstat: 0 Tests: 2 Failed: 1)
  Failed test:  2
  Parse errors: No plan found in TAP output
Files=1, Tests=2,  1 wallclock secs ( 0.01 usr  0.00 sys +  0.15 cusr  0.41 csys =  0.57 CPU)
Result: FAIL
```

**Comment by grimmySwe** [comment]
It seems like the first push for a new branch will not push the entire state, it will only push the already known parts. Up to the .gitrepo commit hash.

**Comment by grimmySwe** [comment]
@ingydotnet I created a new branch issue/106 and implemented a possible solution for this issue.

Flow:
- If it's a push into a non-existing branch, perform initial push with data from .gitrepo:commit and then perform an additional fetch.
- Find common ancestor, in case we already are at front, return
- Perform regular push logic with common ancestor

Created a new test case subrepo-push-new-branch and renamed push-after-init to subrepo-push-after-init.
I do realize that the naming convention of adding subrepo- first on all test are a little silly as we only test subrepo... so a suggestion would be to drop the subrepo prefix for tests and in case we test something else, use prefix for that.

**Comment by ingydotnet** [comment]
Fixed by #110
